uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+    newOffset += sizeof(uint32_t);
+    AssertInBounds(newOffset);
+    char* begin =
+        reinterpret_cast<char*>(&mBuffer.Elements()[newOffset.value()]);
+
+    // Next, ensure the buffer has space for enough bytes up to the size hint.
+    // We may write fewer bytes than hinted, but we need to handle the worst
+    // case if needed.
+    newOffset += aSizeHint;
+    AssertInBounds(newOffset);
+
+    // Call the closure to write the bytes directly into the buffer.
+    size_t bytesWritten = aClosure(Span(begin, aSizeHint));
+
+    // Great, now we know the real length! Write it at the beginning.
+    uint32_t lengthPrefix = PR_htonl(bytesWritten);
+    memcpy(&mBuffer.Elements()[mOffset.value()], &lengthPrefix,
+           sizeof(uint32_t));
+
+    // And figure out our actual offset.
+    newOffset -= aSizeHint;
+    newOffset += bytesWritten;
+    AssertInBounds(newOffset);
+    mOffset = newOffset;
+  }
+
+  RustBuffer ToRustBuffer() {
+    auto size = static_cast<uint32_t>(mOffset.value());
+    auto buffer = {{ ci.ffi_bytebuffer_alloc().name() }}(size);
+    memcpy(buffer.mData, mBuffer.Elements(), size);
+    return buffer;
+  }
+
+ private:
+  void AssertInBounds(const CheckedInt<size_t>& aNewOffset) const {
+    MOZ_RELEASE_ASSERT(aNewOffset.isValid() &&
+                       aNewOffset.value() <= mBuffer.Capacity());
+  }
+
+  template <typename T>
+  void WriteAt(const T& aValue,
+               const std::function<void(size_t, const T&)>& aClosure) {
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    mBuffer.SetLength(newOffset.value());
+    aClosure(mOffset.value(), aValue);
+    mOffset = newOffset;
+  }
+
+  nsTArray<uint8_t> mBuffer;
+  CheckedInt<size_t> mOffset;
+};
+
+/// A "trait" struct with specializations for types that can be read and
+/// written into a byte buffer. This struct is specialized for all serializable
+/// types.
+template <typename T>
+struct Serializable {
+  /// Returns the size of the serialized value, in bytes. This is used to
+  /// calculate the allocation size for the Rust byte buffer.
+  static size_t Size(const T& aValue) = delete;
+
+  /// Reads a value of type `T` from a byte buffer.
+  static bool ReadFrom(Reader& aReader, T& aValue) = delete;
+
+  /// Writes a value of type `T` into a byte buffer.
+  static void WriteInto(Writer& aWriter, const T& aValue) = delete;
+};
+
+/// A "trait" with specializations for types that can be transferred back and
+/// forth over the FFI. This is analogous to the Rust trait of the same name.
+/// As above, this gives us compile-time type checking for type pairs. If
+/// `ViaFfi<T, U>::Lift(U, T)` compiles, we know that a value of type `U` from
+/// the FFI can be lifted into a value of type `T`.
+template <typename T, typename FfiType>
+struct ViaFfi {
+  static bool Lift(const FfiType& aLowered, T& aLifted) = delete;
+  static FfiType Lower(const T& aLifted) = delete;
```

Comment:
This returns `FfiType` instead of using an out param because we call `ViaFfi<T, U>::Lower(T)` for each argument in the argument list when calling an FFI function. An out param would mean we need to allocate some temporary stack variables to hold them, and I’ve nooo idea how to express all that in Askama 🤯 Fortunately, FFI types are all primitives, so there’s no fancy copy constructor or move shenanigans going on here, as would be the case with `Lift`.

Reply:
Another good thing to put in a comment.